parser_v2

jxt1009 opened the pull request:
A different approach at the structure of the new nodes, appears to work a lot better.
Also solved issues with reassignment not working. Everything appears to pass as expected now
The output of the json data looks great too.

jxt1009 committed:
Began to implement the statement list changes

jxt1009 committed:
Began updating the operators with the relational operatations.

jxt1009 committed:
Began work on the b_statement and b_statement_list parse nodes

jxt1009 committed:
Almost completed the for node implementation

jxt1009 committed:
Operator updated to handle the new expressions

jxt1009 committed:
Implemented the while node structure, though the b_stmt class will need
to be updated. It currently doesn't accept any further scopes inside of
each block, and it needs to for if statements and such.

jxt1009 committed:
r_asmt is mostly implemented, still trying to figure out how to
compare the two types on each side. Getting the type of the
RHS is easy, getting the type on the LHS appears to be a lot harder.

jxt1009 committed:
Updated stmt to understand the new tokens and create the respective
nodes

jxt1009 committed:
Created if node and made it also handle the else case, so stmt.c
was updated to move the functionality to the other file

jxt1009 committed:
Updated i_expr to follow the more common standard of returning at the
ending char to check it in the calling function

jxt1009 committed:
Took another node implementation and created the for node from it

jxt1009 committed:
Mostly correct, but needs some tweaking still. The b_stmt currently
doesn't accept any new bodies defined inside of an existing body.
Will be updated shortly to accomodate the correct implementation

jxt1009 committed:
Updated the makelists with the new files

jxt1009 committed:
Updated for loop to match correct context

jxt1009 committed:
Fixed an issue regarding if nodes not being able to be created inside a
body. This has been mostly fixed, though it can only go one layer deep
inside of a body. Something like 'while { if { if {' would likely not work.
This might require a rework of the if node.

 The issue comes in because an if node calls to create a b_stmt_list,
 which makes a b_stmt. Then if while the b_stmt is processing and it
 comes across an if node, it becomes cyclical and couldn't seem to get
 a reliable way to concatenate them at the right spots.

jxt1009 committed:
Updated all nodes to match a more seperated structure. The syntax is
fully verified before making the node, only the final data
is passed to them. This allows recursive tree building and the values
are just filled into the nodes afterwards.

jxt1009 committed:
Fixed issues existing with the reassignment node, appears to work correctly now.

Reed-Terdal commented on Parser/Tree_Nodes/Parse_Nodes/for_node.c:
```diff
@@ -0,0 +1,79 @@
+    retVal->incrementer = incrementer;
+    retVal->body = body;
+    return retVal;
+}
```
What is going on here? Shouldn't the constructor take in the token stream, index and next pointer like every other constructor? If you need to deviate from the pattern, create a helper that the constructor calls, like in i_expr

Reed-Terdal commented on Parser/Tree_Nodes/Parse_Nodes/for_node.c:
```diff
@@ -0,0 +1,79 @@
+GString *for_node_to_json(for_node *forNode) {
+    GString *retval = g_string_new(NULL);
+
+    if (retval != NULL) {
```
checking forNode would be more useful

Reed-Terdal commented on Parser/Tree_Nodes/Parse_Nodes/for_node.h:
```diff
@@ -0,0 +1,54 @@
+ * @param index The index in the token stream to start building an assignment from.
+ * @param next[OUT] A pointer to the index of the next token not part of the assignment.
+ * @return A for loop held in a node
+ */
```
Copy Pasta?

Reed-Terdal commented on Parser/Tree_Nodes/Parse_Nodes/if_node.c:
```diff
@@ -0,0 +1,77 @@
+    retVal->b_true = true_branch;
+    retVal->b_false = false_branch;
+    return retVal;
+}
```
This also does not follow the contsturctor pattern

Reed-Terdal commented on Parser/Tree_Nodes/Parse_Nodes/while_node.c:
```diff
@@ -0,0 +1,83 @@
+    retVal->conditional = conditional;
+    //Return the value
+    return retVal;
+}
```
This also does not follow the constructor pattern

Reed-Terdal commented on Parser/Tree_Nodes/Parse_Nodes/while_node.h:
```diff
@@ -0,0 +1,46 @@
+ * @param value[IN] A token of the condition to continue the loop
+ * @param sign[IN] A statement list to run while the condition is true
+ * @return The new while_node
+ */
```
This doesn't match anything

Reed-Terdal commented on Parser/Tree_Nodes/b_stmt.c:
```diff
@@ -0,0 +1,335 @@
+            Type id_type;
+            if (findIDType(curToken->data, &id_type)) {
+                switch (id_type) {
+                    case jf_void:
```
no check for jf_string, what about concat or charat?

Reed-Terdal commented on Parser/Tree_Nodes/b_stmt.c:
```diff
@@ -0,0 +1,335 @@
+                fprintf(stderr, "Syntax Error: Missing } at end of while statement");
+                exit(-1);
+            }
+
```
All of this should be in while_node.c

Reed-Terdal commented on Parser/Tree_Nodes/b_stmt.c:
```diff
@@ -0,0 +1,335 @@
+            } else {
+                fprintf(stderr, "Syntax Error: Missing } at end of for statement");
+                exit(-1);
+            }
```
All of this should be in for_node.c

Reed-Terdal commented on Parser/Tree_Nodes/b_stmt.c:
```diff
@@ -0,0 +1,335 @@
+            } else {
+                fprintf(stderr, "Syntax Error: Missing } at end of if statement");
+                exit(-1);
+            }
```
All of this should be in if_node.c

Reed-Terdal commented on Parser/Tree_Nodes/i_expr.c:
```diff
@@ -174,7 +183,7 @@ i_expr * internal_i_expr_constructor(i_expr * parent, GArray * token_stream, uns
             curToken = &g_array_index(token_stream, Token, curIndex);
         }
-        if (curToken->type == t_end_paren || curToken->type == t_end_stmt || curToken->type == t_comma)
+        if (nextToken->type == t_end_paren || nextToken->type == t_end_stmt || nextToken->type == t_comma)
```
Why was this changed from curToken to nextToken?

Reed-Terdal commented on Parser/Tree_Nodes/r_asmt.c:
```diff
@@ -0,0 +1,96 @@
+
+    r_asmt *new_asmt = calloc(1, sizeof(r_asmt));
+
+    // Need to try expression before we add the ID, because it cannot be self-referencing
```
stale comment? We should not need to add any IDs when reassigning

Reed-Terdal commented on Parser/Tree_Nodes/r_asmt.c:
```diff
@@ -0,0 +1,96 @@
+    // Need to try expression before we add the ID, because it cannot be self-referencing
+    new_asmt->expression = create_expr(tokenStream, index + 2, next);
+
+    Type *varType = malloc(sizeof(Type));
```
Unneeded and causes a memory leak, you can just create a `Type` variable and reference it in the function call, that way it will be removed from the stack frame when the function returns

Reed-Terdal commented on Parser/Tree_Nodes/r_asmt.c:
```diff
@@ -0,0 +1,96 @@
+    Token *curToken = &g_array_index(tokenStream, Token, index);
+
+    // Does the ID exist, because we cant reassign a non existant assignment
+    if (!findIDType(curToken->data, varType)) {
```
Need to use varType to confirm that the expression type matches the ID type

Reed-Terdal commented on Parser/Tree_Nodes/stmt.c:
```diff
@@ -93,9 +101,16 @@ stmt * create_stmt(GArray * token_stream, unsigned long index, unsigned long * n
+        case t_while:
+        case t_if:
+        case t_for:
+            new_statement->bStmtList = create_b_stmt_list(token_stream, index, next);
```
These should call the individual constructors for each node, not a b_stmt_list.

Reed-Terdal commented on Parser/Tree_Nodes/stmt.c:
```diff
@@ -63,15 +64,22 @@ stmt * create_stmt(GArray * token_stream, unsigned long index, unsigned long * n
+                            new_statement->re_asmt = create_r_asmt(token_stream, index, next);
+                        } else {
+                            new_statement->expression = create_expr(token_stream, index, next);
+                        }
```
Not sure what is going on here. If the first token is an ID then it MUST be a reassign. If the first token is a type ("Integer", "Double", "String") then it is an assignment. Do the checking for ID existaning in the reasssignment constructor.

Reed-Terdal commented on Parser/Tree_Nodes/stmt.h:
```diff
@@ -28,8 +28,10 @@ typedef struct stmt
     expr * expression;
-}stmt;
+    b_stmt_list *bStmtList;
+} stmt;
```
I don't think we described it in the meeting, but the stmt struct should not have a b_stmt_list, it should have elements for an if_node, for_node, and while_node, like b_stmt. The ONLY difference between the two is the lack of asmt in the b_stmt.

jxt1009 committed:
Updated issues with reassignment and restructured the nodes to
be formatted more universally. Only thing missing is comparison
of doubles and strings, which shouldn't be terrible to add. Not something
that is required and being tested though, so not a worry.

Reed-Terdal commented on Parser/Tree_Nodes/for_node.c:
```diff
@@ -16,20 +16,72 @@
-    retVal->body = body;
-    return retVal;
+    (*next)++;
+    Token *check = &g_array_index(token_stream, Token, *next);
```
You should be using index, not next. Next is really only used after you create something with an unknown length. For example, when creating an assigmnment, we create the expression, then use next to make sure the following token is a semicolon

Reed-Terdal commented on Parser/Tree_Nodes/if_node.c:
```diff
@@ -0,0 +1,134 @@
+    if_node *retVal = NULL;
+    retVal = calloc(1, sizeof(if_node));
+    (*next)++;
+    Token *check = &g_array_index(token_stream, Token, *next);
```
You should be using index, not next. Next is really only used after you create something with an unknown length. For example, when creating an assigmnment, we create the expression, then use next to make sure the following token is a semicolon

Reed-Terdal commented on Parser/Tree_Nodes/r_asmt.c:
```diff
@@ -24,27 +24,45 @@ r_asmt *create_r_asmt(GArray *tokenStream, unsigned long index, unsigned long *n
     // Get the variable in the reassignment
-    Token *curToken = &g_array_index(tokenStream, Token, index);
-
+    Token *curToken = &g_array_index(tokenStream, Token, (*next));
```
You should be using index, not next. Next is really only used after you create something with an unknown length. For example, when creating an assigmnment, we create the expression, then use next to make sure the following token is a semicolon

Reed-Terdal commented on Parser/Tree_Nodes/r_asmt.c:
```diff
@@ -24,27 +24,45 @@ r_asmt *create_r_asmt(GArray *tokenStream, unsigned long index, unsigned long *n
+            fprintf(stderr, "Syntax Error: Incorrect reassignment, types do not match.");
+            exit(-1);
+        }
+    }
```
This looks good, only thing is I would verify that the re-assignment ends with a semicolon, this is where you would use the `next` pointer.

Reed-Terdal commented on Parser/Tree_Nodes/while_node.c:
```diff
@@ -23,15 +23,57 @@
-    return retVal;
+    (*next)++;
+    // Load next token
+    Token *check = &g_array_index(token_stream, Token, *next);
```
You should be using index, not next. Next is really only used after you create something with an unknown length. For example, when creating an assigmnment, we create the expression, then use next to make sure the following token is a semicolon

Reed-Terdal reviewed (changes requested):
Need to becareful how you are using the next pointer, it should only be used after something of variable size.

jxt1009 committed:
Updated expressions to understand comparisons, and save them in a way
which is easy to evaluate later.

jxt1009 committed:
Finished the implementation of comparisons in the i_expr module

jxt1009 committed:
Added the implementation necessary for the executer to understand
the new commands. Likely needs some tweaks and cases, but considering
it passes every test, it should be good to go for the phase 2 deadline

jxt1009 committed:
Added the implementation necessary for the executer to understand
the new commands. Likely needs some tweaks and cases, but considering
it passes every test, it should be good to go for the phase 2 deadline

jxt1009 committed:
Fixed up some loose ends in the executer

jxt1009 committed:
Missed a semicolon somehow

jxt1009 committed:
Reverted changes to executer

Reed-Terdal commented on Parser/Tree_Nodes/expr.c:
```diff
@@ -47,10 +46,22 @@ expr * create_expr(GArray * token_stream, unsigned long index, unsigned long * n
             exit(-1);
     }
 
+    Token *cur = &g_array_index(token_stream, Token, index + 1);
```
Make sure to account for more complex expressions infront of comparison

jxt1009 committed:
Updated the formatting on b_stmt_list so that it now consumes
specific tokens regarding it's specific definition.
Also made sure all memory is correctly freed
